### PR TITLE
Restore checkpoint in parallel

### PR DESCRIPTION
**Pull request by SergeyTeplyakov**
I was considering adding another configuration option but decided to stay with the same.

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs [hunk 1]
```diff
@@ -379,6 +376,31 @@ private Task<BoolResult> RestoreCheckpointIncrementalAsync(OperationContext cont
...
+                await ParallelAlgorithms.WhenDoneAsync(
+                    _configuration.IncrementalCheckpointDegreeOfParallelism,
+                    context.Token,
+                    action: async (addItem, kvp) =>
+                    {
+                        var key = kvp.Key;
+                        var value = kvp.Value;
+                        await RestoreFileAsync(context, checkpointTargetDirectory, key, value).ThrowIfFailure();
+                    },
+                    items: newCheckpointInfo.ToArray());
```
Comment (jbayardo) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: Question: what happens if one of the parallel restores fails here?

Comment (SergeyTeplyakov) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs]: In this case WhenDoneAsync will fail with aggregateexception.

---
In reply to: [325331241](https://github.com/microsoft/BuildXL/pull/882#discussion_r325331241) [](ancestors = 325331241)

**Comment by ljcollins25** [comment]
>     internal sealed class CheckpointManager

Should we go ahead and add timeouts around Create/Restore checkpoint?

---
Refers to: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs:29 in 7deb6a0. [](commit_id = 7deb6a03c23d257e1e54f7cd92cd2122aad9acfe, deletion_comment = False)

**Comment by ljcollins25** [comment]
>     internal sealed class CheckpointManager

Actually, that might be non-trivial since we need to ensure that no files are locked which might caused failures in subsequent operations.

---
In reply to: [532680719](https://github.com/microsoft/BuildXL/pull/882#issuecomment-532680719) [](ancestors = 532680719)

---
Refers to: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs:29 in 7deb6a0. [](commit_id = 7deb6a03c23d257e1e54f7cd92cd2122aad9acfe, deletion_comment = False)

**Comment by jbayardo** [comment]
>     internal sealed class CheckpointManager

What do  you mean by "no files are locked"? which files?

---
In reply to: [532681435](https://github.com/microsoft/BuildXL/pull/882#issuecomment-532681435) [](ancestors = 532681435,532680719)

---
Refers to: Public/Src/Cache/ContentStore/Distributed/NuCache/CheckpointManager.cs:29 in 7deb6a0. [](commit_id = 7deb6a03c23d257e1e54f7cd92cd2122aad9acfe, deletion_comment = False)

**Pull request merged**